internal/datacoord/clustering_compaction_manager.go

```diff
@@ -0,0 +1,809 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/indexpb"
+	"github.com/milvus-io/milvus/pkg/common"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/logutil"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/metautil"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"github.com/milvus-io/milvus/pkg/util/tsoutil"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
+)
+
+type ClusteringCompactionManager struct {
+	ctx               context.Context
+	meta              *meta
+	allocator         allocator
+	compactionHandler compactionPlanContext
+	scheduler         Scheduler
+	analyzeScheduler  *taskScheduler
+	handler           Handler
+	quit              chan struct{}
+	wg                sync.WaitGroup
+}
+
+const TaskMaxRetryTimes int32 = 3
+
+type ClusteringCompactionSummary struct {
+	state         commonpb.CompactionState
+	executingCnt  int
+	pipeliningCnt int
+	completedCnt  int
+	failedCnt     int
+	timeoutCnt    int
+	initCnt       int
+	analyzingCnt  int
+	analyzedCnt   int
+	indexingCnt   int
+	indexedCnt    int
+	cleanedCnt    int
+}
+
+func newClusteringCompactionManager(
+	ctx context.Context,
+	meta *meta,
+	allocator allocator,
+	compactionHandler compactionPlanContext,
+	analyzeScheduler *taskScheduler,
+	handler Handler,
+) *ClusteringCompactionManager {
+	return &ClusteringCompactionManager{
+		ctx:               ctx,
+		meta:              meta,
+		allocator:         allocator,
+		compactionHandler: compactionHandler,
+		analyzeScheduler:  analyzeScheduler,
+		handler:           handler,
+	}
+}
+
+func (t *ClusteringCompactionManager) start() {
+	triggers := t.meta.GetClusteringCompactionTasks()
+	for _, tasks := range triggers {
```

Comment:
This logic should not be included here; it should be handled within the state machine.